ci-operator/step-registry/ipi/conf/azure/osimage/ipi-conf-azure-osimage-commands.sh

```diff
@@ -58,56 +58,65 @@ AZURE_AUTH_TENANT_ID="$(<"${AZURE_AUTH_LOCATION}" jq -r .tenantId)"
 
 # log in with az
 az login --service-principal -u "${AZURE_AUTH_CLIENT_ID}" -p "${AZURE_AUTH_CLIENT_SECRET}" --tenant "${AZURE_AUTH_TENANT_ID}" --output none
-# random select the market-place image
-offer="rh-ocp-worker"
-images=$(az vm image list --all --offer ${offer} --publisher redhat -o tsv --query "[?starts_with(urn,'RedHat:${offer}:${offer}')].urn")
-mapfile -t marketImages <<< "${images}"
-echo "choice market-place image from:" "${marketImages[@]}"
-imageTotal=${#marketImages[@]}
-echo "Image total: ${imageTotal} "
-if [ ${imageTotal} -lt 1 ] ; then
-  echo "Fail to find the market-place image under the ${AZURE_AUTH_LOCATION}"
-  exit 1
+
+#Get imageinfo list
+if [[ "${OS_IMAGE_URN}" != "" ]]; then
+    image=${OS_IMAGE_URN}
+    IFS=':' read -ra imageInfo <<< "${image}"
+else
+    # random select image from azure marketplace with sepcified version
+    # Generation: gen1 or gen2
+    offer="rh-ocp-worker"
+    publisher="RedHat"
+    images=$(az vm image list --all --offer ${offer} -o tsv --query "[?publisher=='${publisher}'] | [?version=='${OS_IMAGE_VERSION}'].urn")
+    mapfile -t marketImages <<< "${images}"
+    echo "choice market-place image from:" "${marketImages[@]}"
+    imageTotal=${#marketImages[@]}
+    echo "Image total: ${imageTotal} "
+    if [ ${imageTotal} -lt 1 ] ; then
+        echo "Fail to find the market-place image under the ${AZURE_AUTH_LOCATION}"
+        exit 1
+    fi
+    selected_image_idx=$((RANDOM % ${imageTotal}))
+    image=${marketImages[selected_image_idx]}
+    echo "$(date -u --rfc-3339=seconds) - Selected image ${image}"
+    IFS=':' read -ra imageInfo <<< "${image}"
 fi
-selected_image_idx=$((RANDOM % ${imageTotal}))
-image=${marketImages[selected_image_idx]}
-echo "$(date -u --rfc-3339=seconds) - Selected image ${image}"
-IFS=':' read -ra imageInfo <<< "${image}"
 echo "imageInfo: " "${imageInfo[@]}"
 
-# create a patch to set osImage for compute
-PATCH="/tmp/install-config-existingworkers-marketimage.yaml.patch"
+# image plan is case-sensitive, make sure that publisher/offer/sku keep the same as plan
+if [[ "${OS_IMAGE_PLAN}" == "WithPurchasePlan" ]] || [[ "${OS_IMAGE_PLAN}" == "" ]]; then
+    imageInfo[0]=$(az vm image show --urn ${image} --query 'plan.publisher' -otsv)
+    imageInfo[1]=$(az vm image show --urn ${image} --query 'plan.product' -otsv)
+    imageInfo[2]=$(az vm image show --urn ${image} --query 'plan.name' -otsv)
+fi
+
+#set plan for os image, default value is WithPurchasePlan
+os_image_plan=""
+if [[ "${OS_IMAGE_PLAN}" != "" ]] && [[ ${ocp_minor_version} -ge 14 ]]; then
```

Comment:
I guess we do not need to judge `ocp_minor_version` here, if someone set OS_IMAGE_PLAN on an unsupported version, that should be a job config issue.